File: 0486-predict-the-winner/0486-predict-the-winner.java
class Solution {

    Integer[][] dp;

    private int solve(int left, int right, int[] nums) {

        if (left == right)
            return nums[left];

        if (dp[left][right] != null)
            return dp[left][right];

        int takeLeft = nums[left] - solve(left + 1, right, nums);
        int takeRight = nums[right] - solve(left, right - 1, nums);

        return dp[left][right] = Math.max(takeLeft, takeRight);
    }

    public boolean predictTheWinner(int[] nums) {

        int n = nums.length;
        dp = new Integer[n][n];

        return solve(0, n - 1, nums) >= 0;
    }
}

File: 0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def predictTheWinner(self, nums: List[int]) -> bool:

        memo = {}

        def solve(left, right):

            if left == right:
                return nums[left]

            if (left, right) in memo:
                return memo[(left, right)]

            take_left = nums[left] - solve(left + 1, right)
            take_right = nums[right] - solve(left, right - 1)

            memo[(left, right)] = max(take_left, take_right)

            return memo[(left, right)]

        return solve(0, len(nums) - 1) >= 0